src/cluster/websocket/WebSocket.h

```diff
@@ -0,0 +1,301 @@
+ * on Zeek's main thread.
+ */
+class WebSocketEventDispatcher {
+public:
+    WebSocketEventDispatcher();
+
+    ~WebSocketEventDispatcher();
+
+    /**
+     * Called shutting down a WebSocket server.
+     */
+    void Terminate();
+
+    /**
+     * Queue the given WebSocket event to be processed on Zeek's main loop.
+     *
+     * @param work The WebSocket event to process.
+     */
+    void QueueForProcessing(WebSocketEvent&& event);
+
+    /**
+     * Send a reply to the given websocket client.
+     *
+     * The dispatcher has an internal thread for serializing
+     * and sending out the event.
+     */
+    void QueueReply(WebSocketReply&& reply);
+
+private:
+    /**
+     * Main processing function of the dispatcher.
+     *
+     * This runs on Zeek's main thread.
+     */
+    void Process(WebSocketEvent&& event);
+
+    void Process(WebSocketOpen& open);
+    void Process(WebSocketSubscribeFinished& fin);
+    void Process(WebSocketMessage& msg);
+    void Process(WebSocketClose& close);
+
+
+    /**
+     * Data structure for tracking WebSocket clients.
+     */
+    struct WebSocketClientEntry {
+        std::string id;
+        std::shared_ptr<WebSocketClient> wsc;
+        std::shared_ptr<zeek::cluster::Backend> backend;
+        uint64_t msg_count = 0;
+        std::list<WebSocketMessage> queue;
+    };
+
+
+    void HandleSubscriptions(WebSocketClientEntry& entry, std::string buf);
+    void HandleEvent(WebSocketClientEntry& entry, std::string buf);
+
+    // Allow access to Process(WebSocketEvent)
+    friend zeek::detail::OnLoopProcess<WebSocketEventDispatcher, WebSocketEvent>;
+
+    // Clients that this dispatcher is tracking.
+    std::map<std::string, WebSocketClientEntry> clients;
+
+    // Connector to the IO loop.
+    zeek::detail::OnLoopProcess<WebSocketEventDispatcher, WebSocketEvent>* onloop = nullptr;
+
+    // Thread replying to clients. Zeek's threading manager takes ownership.
+    ReplyMsgThread* reply_msg_thread = nullptr;
+};
+
+/**
+ * An abstract WebSocket server.
+ */
+class WebSocketServer {
+public:
+    WebSocketServer(std::unique_ptr<WebSocketEventDispatcher> demux) : dispatcher(std::move(demux)) {}
+    virtual ~WebSocketServer() = default;
+
+    /**
+     * Stop this server.
+     */
+    void Terminate() {
+        dispatcher->Terminate();
+
+        DoTerminate();
+    }
+
+    WebSocketEventDispatcher& Dispatcher() { return *dispatcher; }
+
+private:
+    /**
+     * Hook to be implemented when a server is terminated.
+     */
+    virtual void DoTerminate() = 0;
+
+    std::unique_ptr<WebSocketEventDispatcher> dispatcher;
+};
+
+
+/**
+ * TLS configuration for a WebSocket server.
+ */
+struct TLSOptions {
+    bool enable = true;
+    bool enable_peer_verification = false;
+    std::string cert_file;
+    std::string key_file;
+    std::string ca_file;
+    std::string ciphers;
+};
+
+struct ServerOptions {
+    std::string host;
+    int port = -1;
```

Comment:
Nit: Probably by design, but `-1` is not a good default port. Could this be e.g., unsigned and default to `0` (typical value for: pick anything)?